### PR TITLE
added `mse_loss`

### LINKED ISSUE #174: `mse_loss`
## 🚀 Feature

Implement [`torch.nn.functional.mse_loss`](https://pytorch.org/docs/stable/generated/torch.nn.functional.mse_loss.html#torch.nn.functional.mse_loss)

### Motivation

NeMo[ text-to-image model](https://github.com/NVIDIA/NeMo/blob/main/examples/multimodal/text_to_image/imagen/imagen_training.py).

### PR DESCRIPTION
**Pull request by k223kim**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #174 .

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

YES!

### CODE REVIEW

**Comment by k223kim** [comment]
Hi @mruberry! 
Just wanted to follow up the `mse_loss` issue and ask some specific questions regarding the implementation. Before anything, without any `DecorateInfo` in `mse_loss_opinfo`, the current implementation has the following errors:
```
FAILED thunder/tests/test_grad.py::test_vjp_correctness_mse_loss_torch_cpu_float64 - AssertionError: Scalars are not close!
FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_bfloat16 - RuntimeError: "mse_cpu" not implemented for 'BFloat16'
FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float16 - RuntimeError: "mse_backward_cpu_out" not implemented for 'Half'
FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float32 - AssertionError: Tensor-likes are not close!
FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float64 - AssertionError: Tensor-likes are not close!
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_mse_loss_torch_cpu_bfloat16 - RuntimeError: "mse_cpu" not implemented for 'BFloat16'
```
1. My question is, I am not sure if the error is due to my implementation itself, or if PyTorch is not supporting something. In the case of the "not implemented error", I am tempted to resolve these with the following `DecorateInfo`:
```python
test_directives=(
    DecorateInfo(
        pytest.mark.skip,
        "test_core_vs_torch_consistency",
        dtypes=(datatypes.bfloat16,),
        devicetypes=(devices.DeviceType.CPU,),
    ),
    DecorateInfo(
        pytest.mark.skip,
        "test_phantom_grad_vs_torch_consistency",
        dtypes=(datatypes.bfloat16, datatypes.float16,),
        devicetypes=(devices.DeviceType.CPU,),
    ),
),
```
How do I know if PyTorch supports this or not?
2. Regarding other assertion errors, as it is throwing errors in for a dtype of `float64`, I am assuming this is because of my implementation. Any suggestions or comments would be appreciated!

File: thunder/clang/__init__.py [hunk 1]
```diff
@@ -1213,14 +1213,21 @@ def matrix_transpose(a: TensorProxy) -> TensorProxy:
 # TODO: add scalar support
 # TODO: review hasattr pattern
 @clangop()
-def maybe_broadcast(*args):
+def maybe_broadcast(*args, preserve_cpu_scalar_tensors=False):
```
Comment (mruberry) [inline thunder/clang/__init__.py:1216]: nit: annotate the type of `preserve_cpu_scalar_tensors` to be `bool`

File: thunder/core/utils.py [hunk 2]
```diff
@@ -635,6 +635,10 @@ def _reify(x):
 #
 
 
+def is_cpu_scalar_tensor(a: Any) -> bool:
```
Comment (mruberry) [inline thunder/core/utils.py:638]: This is a nice utility function (it could also expect that the input is a tensor and not do the first check, but that's fine, too)

File: thunder/executors/torchex.py [hunk 3]
```diff
@@ -1238,6 +1239,30 @@ def _convolution_transform(
     return convolution(a, weight, bias, stride, padding, dilation, bool(transposed), output_padding, groups)
 
 
+def _mse_loss_backward_impl(
```
Comment (mruberry) [inline thunder/executors/torchex.py:1242]: @IvanYashchuk You may want to review this part?

File: thunder/tests/opinfos.py [hunk 4]
```diff
@@ -7238,6 +7238,58 @@ def nll_loss_error_generator(op, device, dtype=torch.float32, **kwargs):
...
+
+
+mse_loss_opinfo = OpInfo(
+    ltorch.mse_loss,
+    supports_grad=True,
+    sample_input_generator=mse_loss_sample_generator,
+    torch_reference=torch.nn.functional.mse_loss,
+    dtypes=(datatypes.floating,),
+    test_directives=(
+        # NOTE Torch does not support bf16 mse_loss
```
Comment (mruberry) [inline thunder/tests/opinfos.py:7271]: Nice comment

File: thunder/tests/opinfos.py [hunk 5]
```diff
@@ -7238,6 +7238,58 @@ def nll_loss_error_generator(op, device, dtype=torch.float32, **kwargs):
...
+        DecorateInfo(
+            pytest.mark.skip,
+            "test_core_vs_torch_consistency",
+            dtypes=(datatypes.bfloat16,),
+            devicetypes=(devices.DeviceType.CPU,),
+        ),
+        DecorateInfo(
```
Comment (mruberry) [inline thunder/tests/opinfos.py:7278]: Add a comment here describing why these tests are skipped

File: thunder/tests/opinfos.py [hunk 6]
```diff
@@ -7238,6 +7238,58 @@ def nll_loss_error_generator(op, device, dtype=torch.float32, **kwargs):
 nn_ops.append(nll_loss_opinfo)
 
 
+def mse_loss_sample_generator(op, device, dtype, requires_grad, **kwards):
+    make = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    # input_shape, target_shape
+    shapes = (
+        ((2, 16), (2, 16)),
```
Comment (mruberry) [inline thunder/tests/opinfos.py:7246]: What about some examples where the shape of the input and the shape of the target differ?

File: thunder/torch/__init__.py [hunk 7]
```diff
@@ -974,6 +974,13 @@ def view(a: TensorLike, /, *shape) -> TensorLike:
     return reshape(a, shape)
 
 
+@torchsymbol(torch.broadcast_tensors, is_method=True)
+def broadcast_tensors(*inputs):
+    if len(inputs) == 1 and not isinstance(inputs[0], TensorProxy):
```
Comment (mruberry) [inline thunder/torch/__init__.py:979]: What's going on with this check?

File: thunder/torch/__init__.py [hunk 8]
```diff
@@ -3906,6 +3913,73 @@ def nll_loss_backward(
...
+    )
+    utils.check(
+        reduction in ("none", "sum", "mean"),
+        lambda: f'Expected reduction string to be "none", "sum", or "mean", but it is {reduction}.',
+        exception_type=ValueError,
+    )
+
+    # warn broadcasting
+    if a.size() != target.size():
+        warnings.warn(
```
Comment (mruberry) [inline thunder/torch/__init__.py:3937]: Would you elaborate on this? I'm curious to understand what's happening better

File: thunder/torch/__init__.py [hunk 9]
```diff
@@ -3906,6 +3913,73 @@ def nll_loss_backward(
...
+            f"Using a target size {target.size()} that is different to the input size {a.size()}"
+            "This will likely lead to incorrect results due to broadcasting."
+            "Please ensure they have the same size."
+        )
+        a, target = broadcast_tensors(a, target)
+    out = pow(sub(a, target), 2)
```
Comment (mruberry) [inline thunder/torch/__init__.py:3943]: This should be able to use the Python operators, like `out = (a - target) ** 2`, and produce what's expected

File: thunder/torch/__init__.py [hunk 10]
```diff
@@ -3906,6 +3913,73 @@ def nll_loss_backward(
...
+    else:
+        raise ValueError(f"Reduction argument {reduction} to mse_loss is not supported")
+
+
+@torchsymbol("mse_loss_backward", id="mse_loss_backward", is_prim=True)
+def mse_loss_backward(g, a, /, target, reduction):
+    return TensorProxy(like=g, shape=a.shape)
+
+
+def _mse_loss_grad(
```
Comment (mruberry) [inline thunder/torch/__init__.py:3963]: @IvanYashchuk may want to take a look at this, too

**Comment by mruberry** [comment]
> Hi @mruberry! Just wanted to follow up the `mse_loss` issue and ask some specific questions regarding the implementation. Before anything, without any `DecorateInfo` in `mse_loss_opinfo`, the current implementation has the following errors:
> 
> ```
> FAILED thunder/tests/test_grad.py::test_vjp_correctness_mse_loss_torch_cpu_float64 - AssertionError: Scalars are not close!
> FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_bfloat16 - RuntimeError: "mse_cpu" not implemented for 'BFloat16'
> FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float16 - RuntimeError: "mse_backward_cpu_out" not implemented for 'Half'
> FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float32 - AssertionError: Tensor-likes are not close!
> FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float64 - AssertionError: Tensor-likes are not close!
> FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_mse_loss_torch_cpu_bfloat16 - RuntimeError: "mse_cpu" not implemented for 'BFloat16'
> ```
> 
> 1. My question is, I am not sure if the error is due to my implementation itself, or if PyTorch is not supporting something. In the case of the "not implemented error", I am tempted to resolve these with the following `DecorateInfo`:
> 
> ```python
> test_directives=(
>     DecorateInfo(
>         pytest.mark.skip,
>         "test_core_vs_torch_consistency",
>         dtypes=(datatypes.bfloat16,),
>         devicetypes=(devices.DeviceType.CPU,),
>     ),
>     DecorateInfo(
>         pytest.mark.skip,
>         "test_phantom_grad_vs_torch_consistency",
>         dtypes=(datatypes.bfloat16, datatypes.float16,),
>         devicetypes=(devices.DeviceType.CPU,),
>     ),
> ),
> ```
> 
> How do I know if PyTorch supports this or not? 

These skips make a lot of sense. I wouldn't worry about supporting grad by adding `mse_loss_backward` and a custom grad formula transform in this PR. @IvanYashchuk may suggest otherwise, and he's our grad expert, though.

> 2. Regarding other assertion errors, as it is throwing errors in for a dtype of `float64`, I am assuming this is because of my implementation. Any suggestions or comments would be appreciated! I am suspecting the implementation of `_mse_loss_backward_impl` in `executors/torchex.py`is incorrect in some way. But again, I am not sure :( Would appreciate any help!

This may be an interesting follow-up issue for a separate PR that targets optimizing this operation's grad formula, but I think this issue may disappear on this PR if this PR doesn't add the grad formula.

**Comment by mruberry** [comment]
Cool PR, @k223kim!

Let's take this PR out of "draft," — I think it's ready for review!

I added some inline comments, particularly about understanding what's happening with the broadcasting better. I'm curious to hear more about that!

I suggest remove the custom `grad` formula from this PR, and seeing if that simplifies the testing. A separate PR can optimize the gradient computation.

File: thunder/tests/opinfos.py (hunk 6 above)
Comment (k223kim) [reply thunder/tests/opinfos.py:7246]: Ah, sure! From the warning in PyTorch shown below, I thought having different shaped input and target is an undesirable usage of `mse_loss` so didn't add a test case for that.
```
            "This will likely lead to incorrect results due to broadcasting."
            "Please ensure they have the same size."
```
However, I do think it is something good to have :)

File: thunder/torch/__init__.py (hunk 8 above)
Comment (k223kim) [reply thunder/torch/__init__.py:3937]: Absolutely! Well, I thought we have to identify when broadcasting occurs. That is because, the `mse_loss` can result in incorrect results due to broadcasting. To simply identify broadcasting, I thought of comparing the sizes of tensors would be sufficient. (referring most to [this implementation](https://github.com/pytorch/pytorch/blob/6054789874629b9740e7ad830d7a3178a86b0af4/torch/nn/functional.py#L3356))

File: thunder/torch/__init__.py (hunk 7 above)
Comment (k223kim) [reply thunder/torch/__init__.py:979]: I think I should remove this.
When referring to [this](https://github.com/pytorch/pytorch/blob/ea61c9cb299b6dfebc57dc9d8821c34321d568ab/torch/_inductor/decomposition.py#L661), I thought there can be a use case when input and target tensors are passed in a list. However, when trying the following, I guess `torch.broadcast_tensors` does not support that. I should have checked this earlier. Sorry for the confusion!
```python
x = torch.randn(1, 2, 3)
y = torch.randn(4, 5, 6)
a, b = torch.broadcast_tensors([x, y])
```
```
  File "/opt/homebrew/anaconda3/lib/python3.11/site-packages/torch/functional.py", line 76, in broadcast_tensors
    return _VF.broadcast_tensors(tensors)  # type: ignore[attr-defined]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: expected Tensor as element 0 in argument 0, but got list
```

**Comment by k223kim** [comment]
Hey @mruberry! Thanks so much for such a detailed review! It'd be awesome if we can have a separate PR for the grad formula.

Just wanted clarify some points regarding your comments.
>  I wouldn't worry about supporting grad by adding mse_loss_backward and a custom grad formula transform in this PR.
> This may be an interesting follow-up issue for a separate PR that targets optimizing this operation's grad formula, but I think this issue may disappear on this PR if this PR doesn't add the grad formula.

Does that mean I should do the following?
- add `mse_loss` to `op_skip` in `tests/test_grad.py`
- remove the following in `torchex.py` (along with the implementation of `mse_loss_backward_impl`)
```python
mse_loss_backward = ex.register_operator(
    "torch_mse_loss_backward_impl", meta=ltorch.mse_loss_backward, fn=_mse_loss_backward_impl
)
_register_implementation(ltorch.mse_loss_backward, mse_loss_backward, checker=_always_executable)
```
- remove `supports_grad=True` in `mse_loss_opinfo` in `opinfos.py`
- remove `_mse_loss_grad` and `mse_loss_backward` in `torch/__init__.py`.
- remove "mse_loss" in `vjp_op_force` in `test_grad.py`

Just wanted to make sure I understood properly :)
With performing all bullet points mentioned above, now I don't have errors regarding `test_grad.py`. (If that is what you were talking about)
I have pushed a comment with the above lines commented out. I can revert/modify after you confirm the changes!

Afterwards, I guess I can open a new PR regarding the grad formulas that were all removed above and ask Ivan for a review?

File: thunder/torch/__init__.py (hunk 7 above)
Comment (mruberry) [reply thunder/torch/__init__.py]: Removing this sounds good. You're right I don't think we or PyTorch support passing tensors in a list. Although someone could do `*lst` if they wanted to pass the tensors in a list to the operation

File: thunder/torch/__init__.py (hunk 8 above)
Comment (mruberry) [reply thunder/torch/__init__.py:3937]: If PyTorch supports this and it doesn't warn then I think we should preserve that behavior in this PR. If we think the behavior is incorrect or dangerous then we can file a new issue to discuss changing it. Let's keep this PR focused on implementing the operation consistent with PyTorch's implementation.

File: thunder/torch/__init__.py [hunk 11]
```diff
@@ -3907,6 +3913,73 @@ def nll_loss_backward(
...
+
+    # warn broadcasting
+    if a.size() != target.size():
+        warnings.warn(
+            f"Using a target size {target.size()} that is different to the input size {a.size()}"
+            "This will likely lead to incorrect results due to broadcasting."
+            "Please ensure they have the same size."
+        )
+        a, target = broadcast_tensors(a, target)  # do I need this?
+    out = (a - target) ** 2
```
Comment (mruberry) [inline thunder/torch/__init__.py:3943]: I think your analysis is correct that because `sub` implicitly broadcasts the tensors to a common shape, there does not need to be an explicit broadcast before it

File: thunder/torch/__init__.py (hunk 8 above)
Comment (k223kim) [reply thunder/torch/__init__.py:3937]: Yup! Sounds good. Just as a note, PyTorch does output a warning when the input tensor and target tensor's sizes are different. Therefore, let me keep the warning and remove `broadcast_tensors`

**Comment by k223kim** [comment]
@IvanYashchuk Thanks so much for the detailed explanation! I just wanted make sure couple of things:

So, with what I have commented out currently, it passes all tests. However, when I set `supports_grad=True` in `mse_loss_opinfo`, it throws the following error:
```
FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float32 - RuntimeError: Trying to flatten t1 = prims.pow(t0, 2.0)  # t1: "cpu f32[2, 16]" to create a grad formula, but it has no subsymbols
FAILED thunder/tests/test_grad.py::test_phantom_grad_vs_torch_consistency_mse_loss_torch_cpu_float64 - RuntimeError: Trying to flatten t1 = prims.pow(t0, 2.0)  # t1: "cpu f64[2, 16]" to create a grad formula, but it has no subsymbols
```
Shouldn't this test pass if thunder automatically creates a backward function? Let me know if I misunderstood anything. Or should `supports_grad` be False?

Also, does that mean I should remove `mse_loss_backward` registeration in `torchex.py` and `mse_loss_backward` in `torch/__init__.py`?
(cc. @mruberry )

**Comment by IvanYashchuk** [comment]
Yes, please remove mse_loss_backward registrations for this PR.

> So, with what I have commented out currently, it passes all tests. However, when I set supports_grad=True in mse_loss_opinfo, it throws the following error

Historically we've had two different implementations for the backward pass and `supports_grad=True` enables testing of one of them, the other one is tested in "vjp" tests in test_grad.py. This may be confusing of course, but let's set `supports_grad=False` here.

File: thunder/clang/__init__.py [hunk 12]
```diff
@@ -1213,14 +1213,21 @@ def matrix_transpose(a: TensorProxy) -> TensorProxy:
 # TODO: add scalar support
 # TODO: review hasattr pattern
 @clangop()
-def maybe_broadcast(*args):
+def maybe_broadcast(*args, preserve_cpu_scalar_tensors: bool = False):
```
Comment (mruberry) [inline thunder/clang/__init__.py:1216]: Let's revert this changes and leave `maybe_broadcast` unchanged in this PR

File: thunder/core/utils.py [hunk 13]
```diff
@@ -635,6 +635,10 @@ def _reify(x):
 #
 
 
+def is_cpu_scalar_tensor(a: TensorProxy) -> bool:
```
Comment (mruberry) [inline thunder/core/utils.py:638]: I think we can remove this change from this PR now

File: thunder/executors/torchex.py [hunk 14]
```diff
@@ -1329,6 +1330,30 @@ def _convolution_transform(
     return convolution(a, weight, bias, stride, padding, dilation, bool(transposed), output_padding, groups)
 
 
+# def _mse_loss_backward_impl(
```
Comment (mruberry) [inline thunder/executors/torchex.py:1333]: The PR can delete this comment

File: thunder/tests/opinfos.py [hunk 15]
```diff
@@ -7316,6 +7316,63 @@ def nll_loss_error_generator(op, device, dtype=torch.float32, **kwargs):
...
+        ((7, 18), (7, 18)),
+        ((3, 4, 2, 3), (3, 4, 2, 3)),
+        ((3, 4, 2, 3), (4, 1, 3)),
```
Comment (mruberry) [inline thunder/tests/opinfos.py:7327]: Nice additional samples

File: thunder/torch/__init__.py [hunk 16]
```diff
@@ -3907,6 +3907,52 @@ def nll_loss_backward(
...
+        return out
+    elif reduction == "sum":
+        return sum(out)
+    elif reduction == "mean":
+        return mean(out)
+    else:
+        raise ValueError(f"Reduction argument {reduction} to mse_loss is not supported")
+
+
+# @torchsymbol("mse_loss_backward", id="mse_loss_backward", is_prim=True)
```
Comment (mruberry) [inline thunder/torch/__init__.py:3951]: This comment can be deleted from the PR, too

**Review by mruberry — COMMENTED** [review]
This looks really good, @k223kim. There are mostly clean-up changes that should be simple to make. There is a test that's failing:

```
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_mse_loss_nvfuser_cuda_float16 - AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 32 (3.1%)
Greatest absolute difference: 0.00048828125 at index (1, 9) (up to 1e-05 allowed)
Greatest relative difference: 0.0014553070068359375 at index (1, 9) (up to 0.001 allowed)
```

and that's OK, the deviation is pretty small. This PR just needs to update the OpInfo to account for this by increasing that test's tolerance. An example of how this is done can be found here:


https://github.com/Lightning-AI/lightning-thunder/blob/e0ab64867a5be914d0548c195a3f850a76c8c397/thunder/tests/opinfos.py#L4394

Essentially, a decorator is specified in the OpInfo and it is automatically added to the generated test. In this case it looks like the decorator needs to set

```
 custom_comparator(partial(assert_close, atol=1e-3, rtol=1e-2)),
```

If that doesn't work we can increase the test tolerance even more. Let me know if you have any questions about this!

**Comment by mruberry** [comment]
> Hi @mruberry! Thanks for the detailed comments :) it's helping me a lot! I can add the following to `opinfos.py`
> 
> ```python
> DecorateInfo(
>     custom_comparator(partial(assert_close, atol=1e-3, rtol=1e-2)),
>     executors=("nvfuser",),
>     dtypes=(datatypes.float16,),
> )
> ```
> 
> However, my concern is that, I am not able to reproduce the error that you have shown above. Would this be an issue? (on my end, all tests are passing..)

I understand, and that's OK, we want to see the CI passing, and the CI and your local hardware may have some differences that cause a small precision difference.

**Comment by k223kim** [comment]
Hey @mruberry !
One last question, regarding the custom `grad` formula for `mse_loss`, what would be the next steps? Should I submit a separate PR regarding that? Or is there a better way to work on it?
(cc. @IvanYashchuk )

**Comment by mruberry** [comment]
> Hey @mruberry ! One last question, regarding the custom `grad` formula for `mse_loss`, what would be the next steps? Should I submit a separate PR regarding that? Or is there a better way to work on it? (cc. @IvanYashchuk )

Absolutely a separate PR would be great. The reason to add a custom grad formula would be to improve performance. So it'd be nice if the PR showed performance of mse_loss forward->backward before the PR, and then after the PR, so we can be sure it improves performance.

File: thunder/tests/opinfos.py [hunk 17]
```diff
@@ -7316,6 +7316,69 @@ def nll_loss_error_generator(op, device, dtype=torch.float32, **kwargs):
...
+        DecorateInfo(
+            pytest.mark.skip,
+            "test_phantom_grad_vs_torch_consistency",
+            dtypes=(
+                datatypes.bfloat16,
+                datatypes.float16,
+            ),
+            devicetypes=(devices.DeviceType.CPU,),
+        ),
+        # Sets more permissive atol and rtol precisions for float16 than assert_close's defaults
```
Comment (mruberry) [inline thunder/tests/opinfos.py:7369]: Nice comment

**Review by mruberry — APPROVED** [review]
Cool! Nice work, @k223kim!

**Comment by mruberry** [comment]
@t-vi @Borda This may require some special push to merge. The test failure is flaky and unrelated
